include/boost/geometry/algorithms/detail/convex_hull/interface.hpp

```diff
@@ -21,86 +21,171 @@
 #ifndef BOOST_GEOMETRY_ALGORITHMS_CONVEX_HULL_INTERFACE_HPP
 #define BOOST_GEOMETRY_ALGORITHMS_CONVEX_HULL_INTERFACE_HPP
 
-#include <boost/array.hpp>
+#include <array>
 
-#include <boost/variant/apply_visitor.hpp>
-#include <boost/variant/static_visitor.hpp>
-#include <boost/variant/variant_fwd.hpp>
-
-#include <boost/geometry/algorithms/detail/as_range.hpp>
 #include <boost/geometry/algorithms/detail/assign_box_corners.hpp>
 #include <boost/geometry/algorithms/detail/convex_hull/graham_andrew.hpp>
+#include <boost/geometry/algorithms/detail/equals/point_point.hpp>
+#include <boost/geometry/algorithms/detail/for_each_range.hpp>
+#include <boost/geometry/algorithms/detail/select_geometry_type.hpp>
+#include <boost/geometry/algorithms/detail/visit.hpp>
 #include <boost/geometry/algorithms/is_empty.hpp>
 
 #include <boost/geometry/core/closure.hpp>
 #include <boost/geometry/core/cs.hpp>
 #include <boost/geometry/core/exterior_ring.hpp>
+#include <boost/geometry/core/geometry_types.hpp>
 #include <boost/geometry/core/point_order.hpp>
 #include <boost/geometry/core/ring_type.hpp>
+#include <boost/geometry/core/tag.hpp>
+#include <boost/geometry/core/tags.hpp>
+#include <boost/geometry/core/visit.hpp>
 
+#include <boost/geometry/geometries/adapted/boost_variant.hpp> // For backward compatibility
 #include <boost/geometry/geometries/concepts/check.hpp>
+#include <boost/geometry/geometries/ring.hpp>
 
-#include <boost/geometry/strategies/convex_hull/services.hpp>
+#include <boost/geometry/strategies/convex_hull/cartesian.hpp>
+#include <boost/geometry/strategies/convex_hull/geographic.hpp>
+#include <boost/geometry/strategies/convex_hull/spherical.hpp>
 #include <boost/geometry/strategies/default_strategy.hpp>
 
 #include <boost/geometry/util/condition.hpp>
+#include <boost/geometry/util/range.hpp>
+#include <boost/geometry/util/sequence.hpp>
+#include <boost/geometry/util/type_traits.hpp>
 
 
 namespace boost { namespace geometry
 {
 
+// TODO: This file is named interface.hpp but the code below is not the interface.
+//       It's the implementation of the algorithm.
 
 #ifndef DOXYGEN_NO_DETAIL
 namespace detail { namespace convex_hull
 {
 
-template <order_selector Order, closure_selector Closure>
-struct hull_insert
+// Abstraction representing ranges/rings of a geometry
```

Comment:
@awulkiew  can you explain the rationale of the `proxy` in more detail?
I glanced through the code but don't get it by that yet.

For example, is it just useful (shorter), or is it essential (we cannot make this improvement without). And why is it necessary for this function `convex_hull` but not (at least not yet) for other algorithms?

Reply:
Short answer:
It's just shorter. It allows to support boxes without rewriting the `graham_andrew` algorithm or calling it twice.

Long answer:
GeometryCollection can store various kinds of geometries, e.g. Rings together with Boxes. We handle Boxes differently in a separate dispatch and generate a ring out of it. So now we have several options:
1. pass GC into `graham_andrew` algorithm and rewrite it to distinguish between various kinds of geometries
2. leave dispatches and `graham_andrew` as they are now and for GC separately calculate a hull for each Box, then the hull for the rest of the GC and then the resulting hull from all of the intermediate hulls
3. leave dispatches and `graham_andrew` as they are now and for GC separately calculate a hull for each Box, but then pass them together with GC into `graham_andrews` transparently, as if the Box hulls were contained in GC instead of Boxes (I've choosen this solution)

As I mentioned in the description this will work only in cartesian. In spherical/geographic we'll have to distinguish between geometries inside the algorithm. But the algorithm will probably be different than `graham_andrews`.

Reply:
Thanks! I'll finish my review next week (assuming there is no hurry)

Reply:
No, take your time, thanks.

Reply:
Oops, this was forgotten. But I'm OK with merging.